src/modalities/loss_functions.py

```diff
@@ -41,3 +41,75 @@ def __call__(self, forward_batch: InferenceResultBatch) -> torch.Tensor:
         # Flatten the tokens
         loss = self.loss_fun(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
         return loss
+
+
+class NCELoss(torch.nn.Module):
+    def __init__(self, asymm: bool, temp: float):
+        """
+        This calculates the noise contrastive estimation loss between embeddings of two different modalities
+        Implementation slightly adapted from https://arxiv.org/pdf/1912.06430.pdf by adding symmeric and asymmetric loss
+
+        Args:
+            asymm (bool): specifies if the loss is calculated in one direction or both directions
+            temp (float): temprature value for regulating loss
+        """
+        super(NCELoss, self).__init__()
+        self.asymm = asymm
+        self.temp = temp
+
+    def forward(self, embedding1: torch.Tensor, embedding2: torch.Tensor, device: torch.device) -> torch.Tensor:
+        """
+        Args:
+            embedding1 (torch.Tensor): embeddings from modality 1 of size batch x dim
+            embedding2 (torch.Tensor): embeddings from modality 2 of size batch x dim
+            device (torch.device): torch device for calculating loss
+
+        Returns:
+            torch.Tensor: loss tensor
+        """
+        # calculating the similarity matrix of size (bxb)
+        sim_matrix = torch.matmul(embedding1, embedding2.t()) / self.temp
+        # numerator of loss: using similarity scores for all positive pairs (e.g., image and its caption)
+        numerator = sim_matrix * torch.eye(sim_matrix.shape[0], device=device)
+        numerator = numerator.sum(dim=0).view(sim_matrix.shape[0], -1)
+        numerator = torch.logsumexp(numerator, dim=1)
+        # denominator of loss: using all similarity scores for all pairs (positive and negative)
+        if self.asymm:
+            denominator = torch.logsumexp(sim_matrix, dim=1)
+        else:
+            denominator = torch.logsumexp(torch.cat((sim_matrix, sim_matrix.permute(1, 0)), dim=1), dim=1)
+        return torch.mean(denominator - numerator)  # calculated in log space
+
+
+class AsymmNCELoss(Loss):
+    def __init__(self, prediction_key1: str, prediction_key2: str, tag: str = "AsymmNCELoss", temp: float = 1.0):
+        """
+        Asymmetric noise contrastive estimation loss
+
+        Args:
+            prediction_key1 (str): key to access embedding 1
+            prediction_key2 (str): key to access embedding 2
+            tag (str, optional): defaults to "AsymmNCELoss".
+            temp (float, optional): tempreture. Defaults to 1.0.
+        """
+        super().__init__(tag)
+        self.prediction_key1 = prediction_key1
+        self.prediction_key2 = prediction_key2
+        self.loss_fun = NCELoss(asymm=True, temp=temp)
```

Comment:
So NCELoss is used as a costume loss and contains the main calculations of contrastive loss and AsymmNCELoss is just more like a wrapper which requires the instance of NCELoss and then inherits from Loss interface. I think they should be kept separate.

Reply:
since the loss does not need to be an `nn.Module`, after an internal discussion, the final change is that the calculation of contrastive loss is packed inside `nce_loss` function, and `NCELoss` now implements `Loss` interface where `is_asymmetric` can be configured in the constructor. For now `nce_loss` is kept inside `loss_functions.py` but if we implement more losses this way, maybe we can consider moving them into another location.